src/main/java/edu/byu/cs/autograder/compile/ModuleIndependenceVerifier.java

```diff
@@ -0,0 +1,59 @@
+package edu.byu.cs.autograder.compile;
+
+import edu.byu.cs.autograder.GradingContext;
+import edu.byu.cs.autograder.GradingException;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ModuleIndependenceVerifier implements StudentCodeInteractor {
+    @Override
+    public void interact(GradingContext context, StudentCodeReader reader) throws GradingException {
+        Set<File> serverFiles = reader.filesMatching(".*server/src/main/java/.*").collect(Collectors.toSet());
+        Path serverPath = context.stageRepo().toPath().resolve("server/src/main/java");
+        Set<String> serverClasses = qualifiedNames(serverFiles, serverPath);
+
+        Set<File> clientFiles = reader.filesMatching(".*client/src/main/java/.*").collect(Collectors.toSet());
+        Path clientPath = context.stageRepo().toPath().resolve("client/src/main/java");
+        Set<String> clientClasses = qualifiedNames(clientFiles, clientPath);
+
+        try {
+            checkImports(context, reader, serverFiles, clientClasses);
+            checkImports(context, reader, clientFiles, serverClasses);
+        } catch (IOException e) {
+            throw new GradingException("Unable to verify module independence", e);
+        }
+    }
+
+    private Set<String> qualifiedNames(Set<File> files, Path relativeTo) {
+        Set<String> classes = new HashSet<>();
+        for(File file : files) {
+            String fileName = relativeTo.relativize(file.toPath()).toString();
+            if(!fileName.contains(".java")) continue;
+            classes.add(fileName.substring(0, fileName.indexOf('.')).replace('/', '.'));
+        }
+        return classes;
+    }
+
+    private void checkImports(GradingContext context, StudentCodeReader reader, Set<File> files, Set<String> classes)
+            throws IOException {
+        for(File file : files) {
+            List<String> contents = reader.getFileContents(file);
+            for (String line : contents) {
+                if (line.matches("(?:abstract )?class|interface|record")) break;
```

Comment:
What about `Enum` and `Annotation` type files?

Reply:
Do you even need the non-capturing, optional match for `(?:abstract )?` clause?

Reply:
How about this: checking until we see anything that doesn't start with "package" or "import"?

Reply:
It makes a ton of sense to have this optimization in place. Defining it is tricky though.

## Existing Approach

Here are several considerations for the existing approach:
* Sometimes there are blank lines between packages and imports, so we need to not let those trigger any early exits.
* There could exist comments in between imports (maybe some imports are commented out instead of removed, or maybe the student added comments for some reason) so we would need to skip those as well.
* I'm not sure if files are allowed to execute static blocks of code, and then import additional statements.
* It feels like a very safe assumption that all import statements precede the first class/record/enum/annotation. Java enforces that only one public object can exist in each file; additionally, it would just be really weird style to author an item, and then import additional items within the same file.

Therefore, it will be safe to keep the `line.matches() ... break;` statement with some modifications to the regex:
* Other items may _partially contain_ these words (i.e. "import package.MyClass;" or "import school.classroom;")

### Working Regex
The following regex will solve address the needs discussed so far. View it and several test cases on [regex101.com](https://regex101.com/r/ZWGoTq/1):
```regex
"^(?!import|\s*//).*\b(class|record|enum|annotation|interface)\b"
```

## Alternative Approach
If desired, another approach could be taken that always scans every line of the file. If the line does not start with "import ", then it is skipped, otherwise, the package name is extracted and compared against the set. This would be simpler, and assumption free. It's possible that the performance difference would never be noticed.

I think the existing approach is sufficient, and the assumptions made are good enough. However, another idea is presented for comparison's sake.

Reply:
Blank lines are already preskipped. I was originally worried about having weird imports, but the I like the negative lookahead